### PR TITLE
feat: text wrapping

### PR DESCRIPTION
**Pull request by ryan-di**
fix https://github.com/excalidraw/excalidraw/issues/7864
fix https://github.com/excalidraw/excalidraw/issues/4267 (partial)

 This PR builds upon the side resizing PR. For text elements, when resized from "left" and "right" (or "w" and "e"), we support text wrapping and unwrapping. 

# follow-ups/TBD

- [ ] drag-with-text-tool during creation to set the initial text width
- [ ] paste of long text should limit the width (disable `autoResize`)
- [x] TBD: on font size change, keep width constant (upon further review it makes more sense from the viewpoint of user intention)

### CODE REVIEW

**Comment by dwelle** [comment]
Some notes as I go along!

- [ ] when binding a previously-wrapped text to a container, it should automatically unwrap it, otherwise it's acting buggy when containerized

File: packages/excalidraw/element/newElement.ts [hunk 1]
```diff
@@ -343,6 +347,9 @@ export const refreshTextDimensions = (
   if (textElement.isDeleted) {
     return;
   }
+  if (textElement.autoResize === false) {
+    return;
+  }
```
Comment (dwelle) [inline packages/excalidraw/element/newElement.ts:352]: I'm not sure about this ad-hoc disabling of this helper based on whether an element has `autoResize` disabled. Adds complexity / surprising behavior. E.g., if you expect something to refresh dimensions of an element, it should at most just not refresh the constrained axis, not bail completely.

File: packages/excalidraw/element/textWysiwyg.tsx [hunk 2]
```diff
@@ -224,6 +225,24 @@ export const textWysiwyg = ({
...
+        const metrics = measureText(
+          text,
+          getFontString(updatedTextElement),
+          updatedTextElement.lineHeight,
+        );
+
+        textElementWidth = updatedTextElement.width;
+        textElementHeight = metrics.height;
+
+        maxHeight = metrics.height;
```
Comment (dwelle) [inline packages/excalidraw/element/textWysiwyg.tsx:245]: similarly to my comment about `refreshTextDimensions`, the fact you disable it there, you now need to handle part of that logic in here, and also start updating `height` & `text` of the text element below, while for `autoResize` text elements we do it upstream.

This leads to having both tight and lose coupling at the same time based on a property. Very hard to follow. Prone to errors I bet as well.

File: packages/excalidraw/actions/actionTextAutoResize.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { isTextElement } from "../element";
+import { newElementWith } from "../element/mutateElement";
+import { measureText } from "../element/textElement";
+import { getSelectedElements } from "../scene";
+import { StoreAction } from "../store";
+import type { AppClassProperties } from "../types";
+import { getFontString } from "../utils";
+import { register } from "./register";
+
+export const actionUnwrapText = register({
```
Comment (ad1992) [inline packages/excalidraw/actions/actionTextAutoResize.ts:10]: can you make the action name same as filename ? so both are in sync and it avoids any confusion as well

File: packages/excalidraw/change.ts [hunk 4]
```diff
@@ -1450,7 +1450,9 @@ export class ElementsChange implements Change<SceneElementsMap> {
         continue;
       }
 
-      redrawTextBoundingBox(boundText, container, elements, false);
+      redrawTextBoundingBox(boundText, container, elements, {
+        informMutation: false,
```
Comment (ad1992) [inline packages/excalidraw/change.ts:1454]: I don't recall, why is this not needed any more ?

File: packages/excalidraw/element/resizeElements.ts [hunk 5]
```diff
@@ -245,17 +248,19 @@ const resizeSingleTextElement = (
...
-  }
-  if (transformHandleType.includes("w")) {
-    scaleX = (x2 - rotatedX) / (x2 - x1);
-  }
-  if (transformHandleType.includes("n")) {
-    scaleY = (y2 - rotatedY) / (y2 - y1);
-  }
-  if (transformHandleType.includes("s")) {
-    scaleY = (rotatedY - y1) / (y2 - y1);
+  if (transformHandleType !== "e" && transformHandleType !== "w") {
```
Comment (ad1992) [inline packages/excalidraw/element/resizeElements.ts:251]: can this go into else of line 327 ? since you are checking if its `e` || `w`

File: packages/excalidraw/element/textElement.ts [hunk 6]
```diff
@@ -62,21 +62,26 @@ export const redrawTextBoundingBox = (
...
   }
+
   const metrics = measureText(
     boundTextUpdates.text,
     getFontString(textElement),
     textElement.lineHeight,
   );
 
-  boundTextUpdates.width = metrics.width;
+  if (textElement.autoResize) {
```
Comment (ad1992) [inline packages/excalidraw/element/textElement.ts:82]:
```suggestion
// Note: for bound text containers, the autoresize will always be set to true
  if (textElement.autoResize) {
```

For bound text containers the attribute will also be true right? In that case lets add a comment

File: packages/excalidraw/element/textWysiwyg.test.tsx [hunk 7]
```diff
@@ -236,6 +236,117 @@ describe("textWysiwyg", () => {
     });
   });
 
+  describe("Test wrapping", () => {
```
Comment (ad1992) [inline packages/excalidraw/element/textWysiwyg.test.tsx:239]:
```suggestion
  describe("Test text wrapping", () => {
```

File: packages/excalidraw/element/textWysiwyg.test.tsx [hunk 8]
```diff
@@ -236,6 +236,117 @@ describe("textWysiwyg", () => {
...
+      updateTextEditor(editor, "");
+      await new Promise((cb) => setTimeout(cb, 0));
+      updateTextEditor(editor, nextText);
+      await new Promise((cb) => setTimeout(cb, 0));
+      editor.blur();
+
+      expect(h.elements[0].width).toEqual(wrappedWidth);
+    });
+
+    it("should restore after unwrapping a wrapped text", async () => {
```
Comment (ad1992) [inline packages/excalidraw/element/textWysiwyg.test.tsx:310]:
```suggestion
    it("should restore original text after unwrapping a wrapped text", async () => {
```

File: packages/excalidraw/element/textWysiwyg.test.tsx [hunk 9]
```diff
@@ -236,6 +236,117 @@ describe("textWysiwyg", () => {
...
+      mouse.doubleClickAt(text.x + text.width / 2, text.y + text.height / 2);
+      updateTextEditor(editor, wrappedText);
+      await new Promise((cb) => setTimeout(cb, 0));
+      editor.blur();
+      // unwrap
+      UI.resize(text, "e", [30, 0]);
+      // expect the text to be restored the same
+      expect((h.elements[0] as ExcalidrawTextElement).text).toBe(prevText);
+    });
+  });
```
Comment (ad1992) [inline packages/excalidraw/element/textWysiwyg.test.tsx:348]:
```suggestion
     const originalText = "Excalidraw\neditor\nis great!";
      const text = API.createElement({
        type: "text",
        text: originalText,
      });
      h.elements = [text];

      // wrap
      UI.resize(text, "e", [-40, 0]);
      // enter text editing mode
      UI.clickTool("selection");
      mouse.doubleClickAt(text.x + text.width / 2, text.y + text.height / 2);
      const editor = await getTextEditor(textEditorSelector);
      editor.blur();
      // restore after unwrapping
      UI.resize(text, "e", [40, 0]);
      expect((h.elements[0] as ExcalidrawTextElement).text).toBe(originalText);

      // wrap again and add a new line
      UI.resize(text, "e", [-30, 0]);
      const wrappedText = text.text;
      UI.clickTool("selection");
      mouse.doubleClickAt(text.x + text.width / 2, text.y + text.height / 2);
      updateTextEditor(editor, `${wrappedText}\nA new line!`);
      await new Promise((cb) => setTimeout(cb, 0));
      editor.blur();
      // remove the newly added line
      UI.clickTool("selection");
      mouse.doubleClickAt(text.x + text.width / 2, text.y + text.height / 2);
      updateTextEditor(editor, wrappedText);
      await new Promise((cb) => setTimeout(cb, 0));
      editor.blur();
      // unwrap
      UI.resize(text, "e", [30, 0]);
      // expect the text to be restored the same
      expect((h.elements[0] as ExcalidrawTextElement).text).toBe(originalText);
    });
  });
```
changed the var name to `originalText` for better readabilty

**Review by ad1992 — COMMENTED** [review]
Hi @ryan-di , did some quick testing of the PR and it looks good ✨
left some comments

File: packages/excalidraw/change.ts (hunk 4 above)
Comment (dwelle) [reply packages/excalidraw/change.ts:1454]: haha, I fell for the same. The `informMutation=false` is kept. Previously it was just a `boolean` flag.

Though through refactor the API signature change is no longer necessary so we can potential revert.

File: packages/excalidraw/actions/actionTextAutoResize.ts (hunk 3 above)
Comment (dwelle) [reply packages/excalidraw/actions/actionTextAutoResize.ts:10]: my bad, forgot to rename action during file rename.

File: packages/excalidraw/tests/resize.test.tsx [hunk 10]
```diff
@@ -426,6 +426,112 @@ describe("text element", () => {
...
+    UI.resize(text, "e", [30, 0]);
+    expect(text.width).toBe(width + 30);
+    expect(text.height).toBe(height);
+
+    UI.resize(text, "e", [-30, 0]);
+    expect(text.width).toBe(width);
+    expect(text.height).toBe(height);
+  });
+
+  it("can be resized from w", async () => {
```
[the remark this answers is not in the document]
Comment (ryan-di) [reply packages/excalidraw/tests/resize.test.tsx:447]: keeping the format: `describe("text element") it("can be resized from w")`

File: packages/excalidraw/change.ts (hunk 4 above)
Comment (ryan-di) [reply packages/excalidraw/change.ts]: reverted 🙂

**Review by dwelle — APPROVED** [review]
